File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.java
import java.util.*;
class Solution {
    public int solution(String[] friends, String[] gifts) {
        int answer = 0;
        int[][] giftList = new int[friends.length][friends.length];
        int[] giftPoint = new int[friends.length];
        HashMap<String,Integer> idMap = new HashMap<>();
        int[] nextGift = new int[friends.length];
        boolean[][] v = new boolean[friends.length][friends.length];
        for(int i = 0; i < friends.length; i++) {
            idMap.put(friends[i],i);
        }
        for(int i = 0; i < gifts.length; i++) {
            String[] str = gifts[i].split(" ");
            int giver = idMap.get(str[0]);
            int receiver = idMap.get(str[1]);
            giftList[giver][receiver]++;
        }
        
        for(int i = 0; i < giftList.length; i++) {
            int give = 0;
            int receive = 0;
            for(int j = 0; j < giftList.length; j++) {
                give += giftList[i][j];
                receive += giftList[j][i];
            }
            giftPoint[i] = give - receive;
        }
        for(int i = 0; i < giftList.length; i++) {
            for(int j = 0; j < giftList.length; j++) {
                if(i == j) continue;
                if(!v[i][j]) {
                    if(giftList[i][j] > giftList[j][i]) 
                        nextGift[i]++;
                    else if(giftList[i][j] < giftList[j][i]) 
                        nextGift[j]++;
                    else {
                        if(giftPoint[i] > giftPoint[j])
                            nextGift[i]++;
                        else if(giftPoint[i] < giftPoint[j])
                            nextGift[j]++;
                    }
                }
                v[i][j] = true;
                v[j][i] = true;
                
            }
        }
        
        Arrays.sort(nextGift);
        answer = nextGift[nextGift.length-1];
        
        
        return answer;
    }
}

File: 프로그래머스/1/258712. 가장 많이 받은 선물/가장 많이 받은 선물.js
function solution(friends, gifts) {
    var answer = 0;
    const give = {};
    const receive = {};
    const giftPoint = {}; // 선물 지수
    const list = {};  // 선물 거래 내역
    const result = {};
    for (const f of friends) {
        give[f] = 0;
        receive[f] = 0;
        list[f] = {};
        result[f] = 0;
    }
       
    for (const gift of gifts) {
        const [A, B] = gift.split(' ');
        give[A]++;
        receive[B]++;
        list[A][B] = (list[A][B] || 0) + 1;
    }
    for (const f of friends) {
        giftPoint[f] = give[f] - receive[f];
    }
    for (const g of friends) {
        for (const r of friends) {
            if (g === r) continue; // 동일인물 제외
            if (list[g][r] && !list[r][g]) {
                result[g]++;
                continue;
            }
            if (!list[g][r] && list[r][g]) {
                result[r]++;
                continue;
            }
            if (list[g][r] && list[r][g]) { // 주고 받은 기록 존재
                if (list[g][r] > list[r][g]) result[g]++;
                else if (list[g][r] < list[r][g]) result[r]++;
                else {
                    if (giftPoint[g] > giftPoint[r]) result[g]++;
                    else if (giftPoint[r] > giftPoint[g]) result[r]++;
                }
            } else {
                if (giftPoint[g] > giftPoint[r]) result[g]++;
                else if (giftPoint[r] > giftPoint[g]) result[r]++;
            }
        }
    }
    answer = Math.max(...Object.values(result)) / 2;
    return answer;
}